### PR TITLE
Parse ServiceConfig JSON string

### PR DESCRIPTION
**Pull request by lyuxuan**

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -788,6 +790,87 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool) (transpor
...
+	Method  *string `json:"method,omitempty"`
+}
+
+func (j jsonName) Stringer() string {
+	res := ""
+	if j.Service != nil {
+		res += "/" + *j.Service + "/"
+	}
+	if j.Method != nil {
+		res += *j.Method
```
Comment (menghanl) [inline clientconn.go]: If `service == nil`, this function returns `"method"`, is this right?

When searching in the map, we search for `/service/method` first, then `/service/`, but never for `method`. Do we need to store the value in the map if `service == nil`?

File: clientconn.go [hunk 2]
```diff
@@ -788,6 +790,87 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool) (transpor
...
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+	var rsc jsonSC
+	err := json.Unmarshal([]byte(js), &rsc)
+	if err != nil {
+		grpclog.Warningf("grpc: handleServiceConfig error unmarshaling %s due to %v", js, err)
+		return err
+	}
+	cc.scRaw = js
+	var sc ServiceConfig
```
Comment (menghanl) [inline clientconn.go]:
```go
sc := ServiceConfig {
￼	LB:      rsc.LoadBalancingPolicy,
￼	Methods: make(map[string]MethodConfig),
}
```

File: clientconn.go [hunk 3]
```diff
@@ -788,6 +790,87 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool) (transpor
...
+}
+
+type jsonSC struct {
+	LoadBalancingPolicy *string   `json:"loadBalancingPolicy,omitempty"`
+	MethodConfig        *[]jsonMC `json:"methodConfig,omitempty"`
+}
+
+// handleServiceConfig parses the service config string in JSON format to Go native
+// struct ServiceConfig, and store both the struct and the JSON string in ClientConn.
+func (cc *ClientConn) handleServiceConfig(js string) error {
```
Comment (menghanl) [inline clientconn.go]: Move the all service config related code (structs and functions) to a new file `service_config.go` and also add unit tests for the json parsing function.

Put all the parsing code in a function
```go
func parseJsonServiceConfig(js string) (ServiceConfig, error)
```

In `ClientConn`, you can do
```go
if sc, err := parse(js); err == nil {
  cc.mu.Lock()
  cc.scRaw = js
  cc.sc = sc
  cc.mu.Unlock()
}
```

File: test/end2end_test.go [hunk 4]
```diff
@@ -1226,104 +1230,192 @@ func newDuration(b time.Duration) (a *time.Duration) {
...
+	te.resolverScheme = r.Scheme()
 	cc := te.clientConn()
+	r.NewAddress([]resolver.Address{{Addr: te.srvAddr}})
+	r.NewServiceConfig(scjs)
 	tc := testpb.NewTestServiceClient(cc)
+
+	// Make sure service config has been processed by grpc.
+	mc := cc.GetMethodConfig("/grpc.testing.TestService/EmptyCall")
+	for {
+		if mc.WaitForReady == nil {
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply test/end2end_test.go]: this code logic has been deleted in recent commit.

File: test/end2end_test.go [hunk 5]
```diff
@@ -1226,104 +1230,192 @@ func newDuration(b time.Duration) (a *time.Duration) {
...
-	}
-	mc2 := grpc.MethodConfig{WaitForReady: newBool(false)}
-	m := make(map[string]grpc.MethodConfig)
-	m["/grpc.testing.TestService/EmptyCall"] = mc1
-	m["/grpc.testing.TestService/"] = mc2
-	sc := grpc.ServiceConfig{
-		Methods: m,
-	}
-	ch <- sc
+	scjs := `{
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply test/end2end_test.go]: done

File: test/end2end_test.go [hunk 6]
```diff
@@ -1226,104 +1230,192 @@ func newDuration(b time.Duration) (a *time.Duration) {
 func TestServiceConfigGetMethodConfig(t *testing.T) {
 	defer leakcheck.Check(t)
 	for _, e := range listTestEnv() {
-		testGetMethodConfig(t, e)
+		if e.balancer == "roundrobin" {
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply test/end2end_test.go]: done

File: clientconn.go (hunk 2 above)
Comment (lyuxuan) [reply clientconn.go]: done

File: clientconn.go (hunk 3 above)
Comment (lyuxuan) [reply clientconn.go]: done

File: clientconn.go (hunk 1 above)
Comment (lyuxuan) [reply clientconn.go]: right. change has been made to skip storing such method config.

File: clientconn.go [hunk 7]
```diff
@@ -816,6 +818,85 @@ func (cc *ClientConn) getTransport(ctx context.Context, opts BalancerGetOptions)
...
+	d, err := time.ParseDuration(*t)
+	return &d, err
+}
+
+type jsonName struct {
+	Service *string `json:"service,omitempty"`
+	Method  *string `json:"method,omitempty"`
+}
+
+func (j jsonName) Stringer() string {
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply clientconn.go]: done

File: service_config.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+type jsonMC struct {
+	Name                    *[]jsonName `json:"name,omitempty"`
+	WaitForReady            *bool       `json:"waitForReady,omitempty"`
+	Timeout                 *string     `json:"timeout,omitempty"`
+	MaxRequestMessageBytes  *int        `json:"maxRequestMessageBytes,omitempty"`
+	MaxResponseMessageBytes *int        `json:"maxResponseMessageBytes,omitempty"`
+}
+
+type jsonSC struct {
```
Comment (menghanl) [inline service_config.go]: This struct is same as the real `ServiceConfig` struct except for the second field. I prefer to not have this temporary struct, and reuse the existing structs.

`json` does a partial parsing if the struct only contains part of the fields, so we could do `Unmarshal` twice, first time for the service/method name, second time for the remaining configs. The second `Unmarshal` can go directly to the existing `MethodConfig` struct.

https://play.golang.org/p/GCSXcP1yQk
```go
const testJSON = `{"name": [ {"service": "foo", "method": "bar"} ], "waitForReady": true}`

type jsonName struct {
	Service *string `json:"service,omitempty"`
	Method  *string `json:"method,omitempty"`
}

type methodConfigName struct {
	Name *[]jsonName
}

type MethodConfig struct {
	WaitForReady *bool
	Timeout      *string
}

func main() {
	buf := []byte(testJSON)

	var mcn methodConfigName
	err := json.Unmarshal(buf, &mcn) // for name only.
	if err != nil {
		panic(err.Error())
	}
	fmt.Printf("%+v\n", mcn)

	var mc MethodConfig
	err = json.Unmarshal(buf, &mc) // for everything other than name.
	if err != nil {
		panic(err.Error())
	}
	fmt.Printf("%+v\n", mc)
}
```

File: clientconn.go [hunk 9]
```diff
@@ -788,6 +789,19 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool) (transpor
 	return t, done, nil
 }
 
+// handleServiceConfig parses the service config string in JSON format to Go native
+// struct ServiceConfig, and store both the struct and the JSON string in ClientConn.
+func (cc *ClientConn) handleServiceConfig(js string) error {
+	sc, err := parseServiceConfig(js)
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply clientconn.go]: done

File: rpc_util.go [hunk 10]
```diff
@@ -524,7 +524,7 @@ type MethodConfig struct {
 type ServiceConfig struct {
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply rpc_util.go]: done

File: service_config.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+	d, err := time.ParseDuration(*t)
+	return &d, err
+}
+
+type jsonName struct {
+	Service *string `json:"service,omitempty"`
+	Method  *string `json:"method,omitempty"`
+}
+
+func (j jsonName) String() (string, bool) {
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply service_config.go]: done

File: service_config.go (hunk 8 above)
Comment (lyuxuan) [reply service_config.go]: filed names are different and go json package cannot handle it.

File: service_config_test.go [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+	}{
+		{
+			`{
+      "loadBalancingPolicy":"round_robin",
+    	"methodConfig": [
+    		{
+    			"name": [
+    				{
+    					"service": "foo",
+              "method": "Bar"
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply service_config_test.go]: done

File: test/end2end_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lyuxuan) [reply test/end2end_test.go]: done

File: service_config_test.go [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc/test/leakcheck"
+)
+
+func TestParseLoadBalancer(t *testing.T) {
+	defer leakcheck.Check(t)
```
[the remark this answers is not in the document]
Comment (lyuxuan) [reply service_config_test.go]: ok to leave it

File: clientconn.go [hunk 14]
```diff
@@ -577,6 +579,7 @@ type ClientConn struct {
 
 	mu    sync.RWMutex
 	sc    ServiceConfig
+	scRaw string
```
Comment (menghanl) [inline clientconn.go]: Set this to `""` in `scWatcher`? Just to be "consistent".

File: service_config_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (menghanl) [reply service_config_test.go]: Still doesn't look right...

**Pull request merged**